libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,607 @@
+        return;
+    }
+
+    _chan = mavlink_channel_t(gcs().get_channel_from_port_number(_mav_port));
+    _mavlink_dynamic_period_ms = 1000; // send dynamic messages at 1 Hz frequency i.e. every 1 second
+    _mavlink_static_period_ms = 10000; // send dynamic messages at 0.1 Hz frequency i.e. every 10 seconds
+}
+
+// Perform the pre-arm checks and prevent arming if they are not satisifed
+// Except in the case of an in-flight reboot
+bool AP_OpenDroneID::pre_arm_check(char* failmsg, uint8_t failmsg_len)
+{
+    WITH_SEMAPHORE(_sem);
+
+    if (!option_set(Options::EnforceArming)) {
+        return true;
+    }
+
+    if (pkt_basic_id.id_type == MAV_ODID_ID_TYPE_NONE) {
+        strncpy(failmsg, "UA_TYPE required in BasicID", failmsg_len);
+        return false;
+    }
+
+    if (pkt_system.operator_latitude == 0 && pkt_system.operator_longitude == 0) {
+        strncpy(failmsg, "operator location must be set", failmsg_len);
+        return false;
+    }
+
+    const uint32_t max_age_ms = 3000;
+    const uint32_t now_ms = AP_HAL::millis();
+
+    if (last_arm_status_ms == 0 || now_ms - last_arm_status_ms > max_age_ms) {
+        strncpy(failmsg, "ARM_STATUS not available", failmsg_len);
+        return false;
+    }
+
+    if (last_system_ms == 0 || now_ms - last_system_ms > max_age_ms) {
+        strncpy(failmsg, "SYSTEM not available", failmsg_len);
+        return false;
+    }
+    
+    if (arm_status.status != MAV_ODID_GOOD_TO_ARM) {
+        strncpy(failmsg, arm_status.error, failmsg_len);
+        return false;
+    }
+    
+    return true;
+}
+
+void AP_OpenDroneID::update()
+{
+    if (_enable == 0) {
+        return;
+    }
+    const uint32_t now = AP_HAL::millis();
+
+    const bool armed = hal.util->get_soft_armed();
+    if (armed && !_was_armed) {
+        // use arm location as takeoff location
+        AP::ahrs().get_location(_takeoff_location);
+    }
+    _was_armed = armed;
+
+    if (now - _last_send_dynamic_messages_ms >= _mavlink_dynamic_period_ms) {
+        _last_send_dynamic_messages_ms = now;
+        send_dynamic_out();
+    }
+
+    send_static_out();
+}
+
+void AP_OpenDroneID::send_dynamic_out()
+{
+    send_location_message();
+}
+
+void AP_OpenDroneID::send_static_out()
+{
+    const uint32_t now_ms = AP_HAL::millis();
+
+    // we need to notify user if we lost the transmitter
+    if (now_ms - last_arm_status_ms > 5000) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "ODID: lost transmitter");
+    }
+
+    const uint32_t msg_spacing_ms = 5000 / 4;
+    if (now_ms - last_msg_send_ms >= msg_spacing_ms) {
+        // allow update of channel during setup, this makes it easy to debug with a GCS
+        _chan = mavlink_channel_t(gcs().get_channel_from_port_number(_mav_port));
+
+        last_msg_send_ms = now_ms;
+        switch (next_msg_to_send) {
+        case NEXT_MSG_BASIC_ID:
+            send_basic_id_message();
+            break;
+        case NEXT_MSG_SYSTEM:
+            send_system_message();
```

Comment:
You now have the system message grouped under the static data. If the GCS is of a type that supports dynamic updates of the operator location (this is [mandatory](https://github.com/opendroneid/opendroneid-core-c#comparison) for integrated drone ID implementations in the US and EU), the operator location will need to be updated once per second in the US or every three seconds for EU . I.e. it is no longer static data.

Reply:
Looks like we pass by the requirement as the `_mavlink_dynamic_period_ms / 4` with this number of messages will get us 1 second? Though how closely we meet that requirement would need to be determined.

And it will be fragile if we add additional messages here.